lightning-invoice/src/payment.rs

```diff
@@ -1113,4 +1213,123 @@ mod tests {
 			}
 		}
 	}
+
+	// *** Full Featured Functional Tests with a Real ChannelManager ***
+	struct ManualRouter(RefCell<VecDeque<Result<Route, LightningError>>>);
+
+	impl<S: routing::Score> Router<S> for ManualRouter {
+		fn find_route(&self, _payer: &PublicKey, _params: &RouteParameters, _first_hops: Option<&[&ChannelDetails]>, _scorer: &S)
+		-> Result<Route, LightningError> {
+			self.0.borrow_mut().pop_front().unwrap()
+		}
+	}
+
+	impl Drop for ManualRouter {
+		fn drop(&mut self) {
+			if std::thread::panicking() {
+				return;
+			}
+			assert!(self.0.borrow_mut().is_empty());
+		}
+	}
+
+	#[test]
+	fn retry_multi_path_single_failed_payment() {
+		// Tests that we can/will retry after a single path of an MPP payment failed immediately
+		let chanmon_cfgs = create_chanmon_cfgs(2);
+		let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+		let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None, None]);
+		let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+
+		create_announced_chan_between_nodes_with_value(&nodes, 0, 1, 1_000_000, 0, InitFeatures::known(), InitFeatures::known());
+		create_announced_chan_between_nodes_with_value(&nodes, 0, 1, 1_000_000, 0, InitFeatures::known(), InitFeatures::known());
+		let chans = nodes[0].node.list_usable_channels();
+		let mut route = Route {
+			paths: vec![
+				vec![RouteHop {
+					pubkey: nodes[1].node.get_our_node_id(),
+					node_features: NodeFeatures::known(),
+					short_channel_id: chans[0].short_channel_id.unwrap(),
+					channel_features: ChannelFeatures::known(),
+					fee_msat: 10_000,
+					cltv_expiry_delta: 100,
+				}],
+				vec![RouteHop {
+					pubkey: nodes[1].node.get_our_node_id(),
+					node_features: NodeFeatures::known(),
+					short_channel_id: chans[1].short_channel_id.unwrap(),
+					channel_features: ChannelFeatures::known(),
+					fee_msat: 100_000_001, // Our default max-HTLC-value is 10% of the channel value
```

Comment:
Is this what makes `ChannelManager` fail the path? Maybe rename `route` to make it explicit why this would fail. That in combination with using the `expect_*` notation would aid readability, IMHO.

Reply:
The `route` object is re-used and made valid later, so it'd be confusing. I started with a method call here that would make it clear but then found that we actually need some further general refactoring for the way we tell the router about the htlc-max our peer set.

Reply:
Sure thing. Though you can always create the valid one and then clone/modify it to get the invalid one with a different name.

I think a bit more clarity in the comment or moving it above the statement would be sufficient, too, for now.